contributors/design-proposals/storage/data-source.md

```diff
@@ -0,0 +1,116 @@
+# Add DataSource for Volume Operations 
+
+Note: this proposal is part of [Volume Snapshot](https://github.com/kubernetes/community/pull/2335) feature design, and also relevant to recently proposed Volume Clone feature. 
+
+## Goal
+In Volume Snapshot proposal, a snapshot is now represented as first-class CRD objects and an external snapshot controller is responsible for managing its lifecycle. With Snapshot API available, users could provision volumes from snapshot and data will be prepopulated to the volumes. Also considering clone and other possible storage operations, there could be many different types of sources used for populating the data to the volumes. In this proposal, we add a general "DataSource" which could be used to represent different types of data sources.
+
+## Design
+A new DataSource field is proposed to add to both PVC and PV to represent the source of the data which is prepopulated to the provisioned volume. If an external-provisioner does not understand the new DataSource field and cannot populate the data to the volume, PV/PVC controller should be able to detect that by comparing DataSource field in PV and PVC (i.e., PVC has DataSource but PV does not) and fail the operation. 
+
+For DataSource, we propose to define a new type “TypedLocalObjectReference”. It is similar to “LocalObjectReference” type with additional Kind field in order to support multiple data source types. In the alpha version, this data source is restricted in the same namespace of the PVC. The following are the APIs we propose to add.
```

Comment:
`support multiple data source types` does this mean `DataSource` could support other sources outside of kubernetes such as Google Cloud buckets or AWS block storage?

Reply:
Theoretically, it is possible since data source allows users to specify different types. But in order to support it, we need to check whether it makes sense and see how to implement it. For example, if your data source is AWS block storage, it does not make sense to try to create a gce disk from this source.